membership/src/main/scala/zio/membership/hyparview/plumtree/PeerState.scala

```diff
@@ -0,0 +1,70 @@
+package zio.membership.hyparview.plumtree
+
+import zio._
+import zio.logging.Logging
+import zio.stm._
+import zio.macros.delegate._
+import zio.membership.hyparview.{ PeerService, TRandom }
+
+trait PeerState[A] {
+  val eagerPushPeers: STM[Nothing, Set[A]]
+  val lazyPushPeers: STM[Nothing, Set[A]]
+  def addToEagerPeers(peer: A): STM[Nothing, Unit]
+  def moveToLazyPeers(peer: A): STM[Nothing, Option[A]]
+  def removePeer(peer: A): STM[Nothing, Unit]
+}
+
+object PeerState {
+
+  final class UsingPartiallyApplied[T] {
+
+    def apply[R <: PeerState[T], E, A](f: PeerState[T] => ZIO[R, E, A]): ZIO[R, E, A] =
+      ZIO.accessM(f(_))
+  }
+
+  def using[T]: UsingPartiallyApplied[T] =
+    new UsingPartiallyApplied[T]
+
+  def withPeerState[A](
+    initialEagerPeers: Int
+  ) =
+    enrichWithM(make[A](initialEagerPeers))
+
+  def make[A](
+    initialEagerPeers: Int
+  ): ZIO[TRandom with PeerService[A] with Logging[String], Nothing, PeerState[A]] = {
+    val initialState = for {
```

Comment:
Move it closed to where it's actually used, that way you can have a nice for comprehension in the end.